### ISSUE TITLE
bounding box format doesn't match for coco dataset

### ISSUE THREAD

**Issue by ruinianxu** [vietnguyen91/Yolo-v2-pytorch#7]
https://github.com/vietnguyen91/Yolo-v2-pytorch/blob/9589413b5dce0476eb9cccc41945cf30cf131b34/src/loss.py#L113

I checked the code for coverting json to pkl and found the bounding box format in annotation file is (xmin, ymin, xmax, ymax). But when matching predicted bbox with groundtruth in buildtarget function, bounding box format seems to be (xmin, ymin, width, height). Am I correct? Or did I miss something?

Thanks

**Comment by HamsterHuey** [comment]
@ruinianxu - It's pretty confusing, but I believe he changes to x_center, y_center, w, h format in one of the transforms that are applied to the dataset. It is not a great way to structure the code and makes it pretty confusing to have it hidden away there, but that is how it is. In general, the model requires/assumes that the final targets it receives should be in x_center, y_center, width, height

**Comment by ruinianxu** [comment]
@HamsterHuey Thanks so much for your reply. I am also confused by scaling groundtruth bbox by reduction factor which is 32. I didn't find this usage in other implementation. Is this in the original work?

**Comment by HamsterHuey** [comment]
Yes. The network reduces input pixel dimension by a factor of 32, so a
416x416 input image ends up at the final layer as a 13 x 13 grid. The
network then tries to predict the co-ordinates of the box within each grid
box, but the overall co-ordinates for the box are defined on this 13 x 13
grid, and that is what the co-ordinate loss is calculated on as well I
believe.

So it is a function of the way YOLO is implemented.

**Comment by HamsterHuey** [comment]
As an addition, see this figure in the YOLO v2 paper:
![image](https://user-images.githubusercontent.com/3207370/57186005-54a0e300-6ea5-11e9-9219-7ea8b0d80b50.png)

It isn't completely clear here, but the co-ordinate system is such that each grid has unit length. This is why the offset of the box center within a box is bound by a sigmoid, so it spans 0 - 1, which lets it span any point within a grid-cell. `c_x` and `c_y` are actually integers, and in the picture as shown, if the top-left corner is 0,0 - `c_x` and `c_y` should both be 1.

**Comment by ruinianxu** [comment]
@HamsterHuey Thank you so much for your reply! I previously considered reduction as another parameter 
 which defines max_bbox_per_image hence I thought it is wired.

**Comment by relotonp** [comment]
@HamsterHuey where exactly is the conversion from [xmin,ymin,xmax,ymax] to [xmin, ymin, width, height] ?

You said that it's at a transformation on the dataset but I can't find it.

The bbox is read out from voc annotation data as [xmin,ymin,xmax,ymax, class_name] [here ](https://github.com/uvipen/Yolo-v2-pytorch/blob/master/src/voc_dataset.py#L41)

Then in line 42/43 multiple transformations are used, but they don't change the format
In [resize ](https://github.com/uvipen/Yolo-v2-pytorch/blob/master/src/data_augmentation.py#L125) it still returns [resized_xmin, resized_ymin, resize_width, resize_height, class_name]

After that we don't use any transformations anymore and in [train_voc ](https://github.com/uvipen/Yolo-v2-pytorch/blob/master/train_voc.py#L109) we get the picture and this annotation as label. Then we [feed ](https://github.com/uvipen/Yolo-v2-pytorch/blob/master/train_voc.py#L116) this label into our loss.

So we have still the format [xmin, ymin, xmax, ymax]
The label(now under the name target) is fed unchanged into [build_target](https://github.com/uvipen/Yolo-v2-pytorch/blob/master/train_voc.py#116)

It is now called ground_truth. 

`            for i, anno in enumerate(ground_truth[b]):
                gt[i, 0] = (anno[0] + anno[2] / 2) / self.reduction
                gt[i, 1] = (anno[1] + anno[3] / 2) / self.reduction
                gt[i, 2] = anno[2] / self.reduction
                gt[i, 3] = anno[3] / self.reduction`

If ground truth is in format [xmin, ymin, xmax. ymax]
anno[2] cannot be width and anno[3] cannot be heigth.

I know that he wants to make gt[i,0] to x_middle and gt[i,1] to y_middle but this calculation is wrong if anno[2] is xmax and not width.
So where did I miss the conversion where he changed the annotation to [xmin,ymin,width,height]?

**Comment by HamsterHuey** [comment]
@relotonp - I believe it happens in the `Resize` transform here: https://github.com/uvipen/Yolo-v2-pytorch/blob/master/src/data_augmentation.py#L125

In general, your best bet would be to switch to looking at the Ultralytics Yolo V3 repo for a decent implementation that has been tested to train to a good mAP value from scratch on COCO. This repo has a lot of bugs, and Yolo in general is a pretty finicky model with lots of crazy implementation details that are easy to get wrong.
